internal/registry/aws/auth.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/google/go-containerregistry/pkg/authn"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/fluxcd/image-reflector-controller/internal/registry"
+)
+
+// ParseImage returns the AWS account ID and region and `true` if
+// the image repository is hosted in AWS's Elastic Container Registry,
+// otherwise empty strings and `false`.
+func ParseImage(image string) (accountId, awsEcrRegion string, ok bool) {
+	registryPartRe := regexp.MustCompile(`([0-9+]*).dkr.ecr.([^/.]*)\.(amazonaws\.com[.cn]*)/([^:]+):?(.*)`)
```

Comment:
Ideally `regexp.MustCompile` would be placed outside of the func, so we can cache that operation. The performance impact:
```
BenchmarkCurrent-16               173902              5948 ns/op            9885 B/op         61 allocs/op
BenchmarkPreCompiled-16          1546942               832.4 ns/op           433 B/op          3 allocs/op
```

Code:
```golang
func BenchmarkCurrent(b *testing.B) {
	for i := 0; i < b.N; i++ {
		registryPartRe := regexp.MustCompile(`([0-9+]*).dkr.ecr.([^/.]*)\.(amazonaws\.com[.cn]*)/([^:]+):?(.*)`)
		_ = registryPartRe.FindAllStringSubmatch("012345678901.dkr.ecr.us-east-1.amazonaws.com/foo:v1", -1)
	}
}

func BenchmarkPreCompiled(b *testing.B) {
	registryPartRe := regexp.MustCompile(`([0-9+]*).dkr.ecr.([^/.]*)\.(amazonaws\.com[.cn]*)/([^:]+):?(.*)`)
	for i := 0; i < b.N; i++ {
		_ = registryPartRe.FindAllStringSubmatch("012345678901.dkr.ecr.us-east-1.amazonaws.com/foo:v1", -1)
	}
}
```

Reply:
Moved the regexp outside, package global.